### PR TITLE
[RFC] making compute CLI command usable inside Python

### PR DESCRIPTION
**Pull request by luizirber**
I'm opening this PR for requesting comments about how to properly support calling CLI commands from the Python API, without being too verbose.

~~This is similar to the design I'm going for in [sourmash-utils][0].~~ (Old PR comments, now this is adapted to #785 ideas)

I'll add inline comments with details, but the rationale is:
We focus on exposing the CLI, but if I want to calculate a signature using the Python API there is quite a bit of boilerplate involved:
```python
import screed
minhashes = []
for g in filenames:
    E = sourmash.MinHash(n=500, ksize=31)
    for record in screed.open(g):
        E.add_sequence(record.sequence, True)
    minhashes.append(E)
```
what I want to do:
```python
from sourmash.command_compute import compute
sigs = compute(filenames)
```
(or something similar).

The idea of this PoC is to  make arguments explicit in the function signature. The CLI exposes commands and options from subparsers, and function arguments are 'fake-positional': only use kwargs, but leave the required ones (like `filenames` in the `compute(filenames)` example) in the beginning to avoid having to define explicitly all the keywords.

## Checklist

- [ ] Is it mergeable?
- [ ] `make test` Did it pass the tests?
- [ ] `make coverage` Is the new code covered?
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

[0]: https://github.com/dib-lab/sourmash-utils

### CODE REVIEW

**Comment by luizirber** [comment]
I talked with@standage yesterday and he has another approach for kevlar, where you build a list of CLI-like arguments and then parse with argparse, passing the resulting `args` object into the main function: https://github.com/dib-lab/kevlar/blob/3130636d5b3822502708084e547b63b671fb2cc2/kevlar/tests/test_dump.py#L36
(and he is also using subparsers:
https://github.com/dib-lab/kevlar/blob/3130636d5b3822502708084e547b63b671fb2cc2/kevlar/cli/__init__.py#L53 )

**Comment by luizirber** [comment]
I talked with @betatim today, and he has a similar setup for `osf-cli`:
https://github.com/dib-lab/osf-cli/blob/b9c750511595cd6b372eae70965a53055fe8933a/osfclient/__main__.py#L23

**Comment by ctb** [comment]
What's old is new!

https://github.com/ctb/twill/blob/master/twill/commands.py

**Comment by luizirber** [comment]
For future ref: https://realpython.com/blog/python/comparing-python-command-line-parsing-libraries-argparse-docopt-click/#commands

**Comment by luizirber** [comment]
I was talking with @taylorreiter and it seems like this is a good PR to start going more into sourmash code, so I updated it to the current master.

After this PR was created there were code changes (for `search` and `gather`) where @ctb took some code away from `sourmash/commands.py` and into function that are easier to call in Python. The issue of exposing the CLI commands as Python functions is that they take A LOT of arguments, but I think it is still good to have them available.

File: sourmash/__main__.py [hunk 1]
```diff
@@ -73,8 +85,19 @@ def main():
...
+        subparsers = parser.add_subparsers()
+        subp = subparsers.add_parser(args.command)
+        commands[args.command].add_args(subp)
+        subp.set_defaults(cmd=commands[args.command])
+        args = parser.parse_args()
+        return args.cmd(**vars(args))
+    else:
+        cmd = commands.get(args.command)
+        return cmd(sys.argv[2:])
+
```
Comment (luizirber) [inline sourmash/__main__.py]: Once all commands are converted we can remove this `if` block and keep only the first part (the `else` is the current implementation for subcommands)

File: sourmash/__main__.py [hunk 2]
```diff
@@ -12,6 +12,13 @@
                        plot, watch, info, storage, migrate)
 from .lca import main as lca_main
 
+try:
+    from sourmash_utils.__main__ import main as utils_main
+    UTILS_AVAILABLE = True
+except ImportError:
+    UTILS_AVAILABLE = False
```
Comment (luizirber) [inline sourmash/__main__.py]: This is here because the original PR was against #236 and I changed it to be against master to have Travis run the tests...

**Comment by ctb** [comment]
I updated in this PR in #722, which resolves the conflicts. Still have some kind of bug in there but it's safe to merge.

**Comment by luizirber** [comment]
@ctb can you take a look at this commit: https://github.com/dib-lab/sourmash/pull/245/commits/193748c87a34822d6c2ac720590df34150fa03d9

This is adding a `ksize_parser` function for the args (that converts strings into a ksize list), and also modifying the cli test to also run with the `compute` function. If it looks OK, I would go modify the other tests (for `sourmash compute`) to match the modified one.

**Comment by luizirber** [comment]
I'm also starting to feel like this should move from `[RFC]` to something actually mergeable.

**Comment by luizirber** [comment]
Revamped this on top of #785, and updated the PR description. Comments @ctb @standage?

**Comment by luizirber** [comment]
> I like everything about this, except for the final command "path" of `sourmash.command_compute.compute`, which seems ...long. No immediate solution (and happy to merge this!) but let's think about how to simplify that!

yeah, the `command_compute` thing is... long.

What I've been doing in the Rust crate is creating submodules over time. Example: at the moment the `src` dir is
```
.
├── cmd.rs
├── errors.rs
├── ffi/
├── from.rs
├── index/
├── lib.rs
├── signature.rs
├── sketch/
└── wasm.rs
```
`ffi/`, `index/` and `sketch/` started as single files (like `cmd.rs`, `signature.rs` and `wasm.rs` are now). As more features were implemented I expand the single file into a submodule, but still using the top-level name (`index`, for example) to control what to export (for the public API).

On the Python side, it would translate to `commands.py` becoming a dir module, and importing/exposing items from its submodules.

Currently:
```
sourmash
├── cli/
├── command_compute.py
├── commands.py
...
└── utils.py
```
Proposed (like what `sourmash.cli` is doing now):
```
sourmash
├── cli/
├── commands/
│   ├── __init__.py
│   └── compute.py
...
└── utils.py
...

This way we can start with a massive implementation file (like `commands.py` is now) and keep the API working when we reorganize into a submodule and split the code (because both could be accessed by `from sourmash.commands import compute`, for example). Doing `sourmash/command_compute.py` goes against this (and I know you like flat structures =]), but it will also start to get hard to find files if all of them are in `sourmash/` (which is similar to the `sourmash/sbt*.py` situation)

Before merging this, I think it is good to have some guidelines on how to proceed for all commands. From the top of my head:
- Required arguments in the CLI -> positional arguments in the `commands.py` function
  * Ideally we could use `*` in the function signature and make some args [keyword-only], but that's not valid in Python 2 =(
- Keeping defaults from argparse in sync with defaults in the `commands.py` function
- Commands taking output (like `sourmash compute -o out input`): how to map `out` into an argument? Accept strings and opened files (`argparse` open a file)? (This is the failing test in this PR, for example)
- docstrings for each command (#828)
- testing: update all tests calling `utils.runscript` to also test the API command? That's what I did with `test_do_sourmash_compute` and `test_do_sourmash_compute_output_valid_file` in this PR.
  * Better: modify `utils.runscript` to do that, if it is passed an extra arg?

[keyword-only]: https://www.asmeurer.com/python3-presentation/slides.html#11

**Comment by ctb** [comment]
@luizirber has thus declared: "#853 should fix the commands-taking-output problem mentioned above." yay w00t!

**Comment by luizirber** [comment]
Still think this is useful, especially considering pyodide #2433 and using it in a browser repl, but lots to update... And so I'm closing it.

**Pull request closed without merging**